client/internal/engine.go

```diff
@@ -857,3 +863,116 @@ func (e *Engine) receiveSignalEvents() {
 
 	e.signal.WaitStreamConnected()
 }
+
+func (e* Engine) parseNATExternalIPMappings() []string {
+	var mappedIPs []string
+	for _, mapping := range e.config.NATExternalIPs {
+		var external, internal string
+		var externalIP, internalIP net.IP
+		var err error
+		split := strings.Split(mapping, "/")
+		if len(split) > 1 {
+			internal = split[1]
+			internalIP = net.ParseIP(internal)
+			if internalIP == nil {
+				// not a properly formattted IP address, maybe it's interface name?
+				internalIP, err = findIPFromInterfaceName(internal)
+				if err != nil {
+					log.Warnf("error finding interface IP for interface '%s', ignoring external mapping '%s': %v", internal, mapping, err)
+					break
+				}
+			}
+		}
+		external = split[0]
+		if external == "stun" {
+			externalIP, err = e.findNATExternalIPFromSTUN(internalIP)
+			if err != nil {
+				log.Warnf("error finding external IP via stun, ignoring external IP mapping '%s': %v", mapping, err)
+				break
+			}
+		} else {
+			externalIP = net.ParseIP(external)
+			if externalIP == nil {
+				log.Warnf("invalid external IP, ignoring external IP mapping '%s'", mapping)
+				break
+			}
+		}
+		if externalIP != nil {
+			mappedIP := externalIP.String()
+			if internalIP != nil {
+				mappedIP = mappedIP + "/" + internalIP.String()
+			}
+			mappedIPs = append(mappedIPs, mappedIP)
+			log.Infof("parsed external IP mapping of '%s' as '%s'", mapping, mappedIP)
+		}
+	}
+	return mappedIPs
+}
+
+func (e *Engine) findNATExternalIPFromSTUN(internalIP net.IP) (net.IP, error) {
+	if len(e.STUNs) < 1 {
+		return nil, errors.New("no STUN servers available for external IP discovery")
+	}
+
+	// What about auth?
+	url := e.STUNs[0]
+	addr :=  net.JoinHostPort(url.Host, strconv.Itoa(url.Port))
+
+	d := net.Dialer{}
+	if internalIP != nil {
+		d.LocalAddr = &net.UDPAddr{ IP: internalIP }
+	}
+	con, err := d.Dial("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+	cli, err := stun.NewClient(con)
+	defer func() {
+		_ = cli.Close()
+	}()
+
+	req := stun.MustBuild(stun.TransactionID, stun.BindingRequest)
+	var hErr *error
+	var hAddr *net.IP 
+	// call to Do method blocks until all handlers are called and have returned, no synchronization needed
+	if err := cli.Do(req, func(e stun.Event) {
+		if e.Error != nil {
+			hErr = &e.Error
+		} else {
+			var xorAddr = stun.XORMappedAddress{}
+			err = xorAddr.GetFrom(e.Message)
+			if err != nil {
+				hErr = &err
+			} else {
+				hAddr = &xorAddr.IP
+			}
+		}
+	}); err != nil {
+		return nil, err
+	}
+	if hErr != nil {
+		return nil, err 
+	}
+	return *hAddr, nil
+}
+
+func findIPFromInterfaceName(ifaceName string) (net.IP, error) {
```

Comment:
We need to check if the interface name has not been blacklisted as the nat map will be ignored if the ip from the interface is not founded. See here: https://github.com/pion/ice/blob/master/gather.go#L164